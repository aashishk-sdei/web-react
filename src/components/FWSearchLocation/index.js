import React, { useRef, useEffect, useCallback } from 'react';
import Typeahead from '../Typeahead';
import { tr } from "../../components/utils";
import { useTranslation } from "react-i18next";
import { useSelector, useDispatch } from 'react-redux';
import { v4 as uuidv4 } from "uuid";
// Actions
import { getAutoCompleteTest } from "../../redux/actions/family";
import { getLocationGUID } from "../../redux/actions/ww1";
import { getApiCancelToken } from "../../redux/requests";
import { getRelatedValue } from "./../../utils"
var timer;
const getSearchText = (val) => (val) ? val : ""

const getLocationGUIDCall = async (id, name, form, props, field) => {
    form.setFieldValue(`${field.name}`, { id: id, name: name })
    if (props.searchType) {
        const data = await getLocationGUID(id);
        if (data) {
            form.setFieldValue(`${field.name}`, { id: id, name: name, "levelData": data })
            if (props.relatedField) {
                const relatedValue = getRelatedValue(form.values.matchExact, data.residenceLevel ? Object.keys(data.residenceLevel)[0] : "")
                form.setFieldValue(`${props.relatedField}`, relatedValue);
            }
        }
    }
}
const cancelPreviousToken = (source) => {
    if (source.current) {
        source.current.cancel();
    }
}
const SearchLocation = ({ field, form, ...props }) => {
    const { t } = useTranslation();
    const { options, optionLoading } = useSelector(state => {
        return state.family
    });
    const source = useRef(null);
    const periousName = useRef(null);
    let requestId = useRef(null);
    const dispatch = useDispatch();
    const callbackApi = useCallback((val) => {
        requestId.current = uuidv4();
        if (timer) {
            clearTimeout(timer);
        }
        const searchText = getSearchText(val)
        timer = setTimeout(() => {
            if (periousName.current !== searchText) {
                periousName.current = searchText
                cancelPreviousToken(source)
                const sourceToken = getApiCancelToken();
                source.current = sourceToken;
                dispatch({ type: "AUTOCOMPLETE_REQUEST" })
                dispatch(getAutoCompleteTest(searchText, sourceToken, requestId.current));
            }
        }, 250)
    }, [dispatch])
    useEffect(() => {
        if (!periousName.current) {
            callbackApi(field?.value?.name || "")
        }
    }, [callbackApi, periousName])
    const loadNextPage = (page) => {
        const sourceToken = getApiCancelToken();
        source.current = sourceToken;
        dispatch({ type: "AUTOCOMPLETE_PAGINATION_REQUEST" })
        dispatch(getAutoCompleteTest(periousName.current, sourceToken, requestId.current, page));
    }
    const handleChange = (val) => {
        if (val?.id) {
            getLocationGUIDCall(val.id, val.name, form, props, field)
        } else {
            form.setFieldValue(`${field.name}`, { "id": "", name: val?.name || "" });
        }
        if (!val?.name) {
            form.setFieldValue(`${props.relatedNameField}`, "1")
        } else {
            if (form?.values?.matchExact) {
                form.setFieldValue(`${props.relatedNameField}`, "0")
            }
        }
    }
    const value = field.value;
    return (
        <div className="flex w-full mt-0">
            <Typeahead
                id={props.id}
                options={options}
                optionLoading={optionLoading}
                handleChange={handleChange}
                pagination={true}
                {...props}
                callbackApi={callbackApi}
                loadNextPage={loadNextPage}
                highlight={props.highlight}
                placeholder={props.placeholder ? props.placeholder : tr(t, "search.unisearchform.autocomplete")}
                getOptionLabel={(option) => {
                    return option.name || ""
                }}
                value={value}
                loading={optionLoading}
            />
        </div>
    );
}

SearchLocation.defaultProps = {
    searchType: false,
    highlight: false
}

export default SearchLocation;