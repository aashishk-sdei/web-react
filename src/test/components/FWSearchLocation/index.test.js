import React from "react";
import { mount } from "enzyme";
import { Field, Formik } from "formik";
import FWSearchLocation from "../../../components/FWSearchLocation";
import {useDispatch, useSelector} from "react-redux";
import { act } from 'react-dom/test-utils';
jest.useFakeTimers();
jest.mock("react-redux", () => ({
    ...jest.requireActual("react-redux"),
    useSelector: jest.fn(),
    useDispatch: ()=>jest.fn()
}));
jest.mock('react-i18next', () => ({
    // this mock makes sure any components using the translate hook can use it without a warning being shown
    useTranslation: () => {
      return {
        t: (str) => str,
      };
    },
}));


describe('Test SearchPeoople component with props', () => {
    let searchPeroopleComponent;
    const mockCallBack = jest.fn(()=> true);
    beforeEach(() => {
        useSelector.mockImplementation(callback => {
            return callback({
                family: {
                    options: [],
                    optionLoaing: true
                }
            });
          });
         searchPeroopleComponent =  mount(<Formik
            initialValues={{ location: {}}}
            enableReinitialize={true}
          >
            <Field
              id="location"
              name="location"
              searchType = {true}
              relatedField= "l.l.s"
              component={FWSearchLocation}
              freeSolo={true}
            />
       </Formik>)
    })
    it('match snapshot', () => {
        expect(searchPeroopleComponent).toMatchSnapshot();
    });
   it('handle Input Change',  async () => {
        useSelector.mockImplementation(callback => {
            return callback({
                family: {
                    options: [
                        {
                            "id": "2a719cc0-6c01-4dea-90c3-5ede17ced68b",
                            "name": "Austria-Hungary (historical)"
                        }
                    ],
                    optionLoaing: false
                }
            })
        });
        const searchInput = searchPeroopleComponent.find("input");
        const autoComplete = searchPeroopleComponent.find('[role="combobox"]');
        await act(async () => {
            const event = { target: {name: "search", value: "Austria"}};
            searchInput.simulate('change', event);
        });
        await act(async () => {
            autoComplete.simulate('keyDown', { key: 'ArrowDown' });
        });
        await act(async () => {
            autoComplete.simulate('keyDown', { key: 'Enter' });
            expect(mockCallBack.mock.calls.length).toEqual(0);
        });
    });
    it('handle Input Change without id',  async () => {
        const searchInput = searchPeroopleComponent.find("Typeahead");
        await act(async () => {
            searchInput.prop('handleChange')({name: "Austria", id: ""});
        });
        await act(async () => {
            jest.runOnlyPendingTimers();
        });
        expect(mockCallBack.mock.calls.length).toEqual(0);
    });
    it('handle Input Change undefined',  async () => {
        const searchInput = searchPeroopleComponent.find("Typeahead");
        await act(async () => {
            searchInput.prop('handleChange')({name: "", id: ""});
        });
        expect(mockCallBack.mock.calls.length).toEqual(0);
    });
    it('handle Input Change with id',  async () => {
        const searchInput = searchPeroopleComponent.find("Typeahead");
        await act(async () => {
            searchInput.prop('handleChange')({
                "id": "2a719cc0-6c01-4dea-90c3-5ede17ced68b",
                "name": "Austria-Hungary (historical)"
            });
        });
        await act(async () => {
            jest.runOnlyPendingTimers();
        });
        expect(mockCallBack.mock.calls.length).toEqual(0);
    });
    it('handle Input Change without loadNextPage',  async () => {
        const searchInput = searchPeroopleComponent.find("Typeahead");
        await act(async () => {
            searchInput.prop('loadNextPage')(2);
        });
        expect(mockCallBack.mock.calls.length).toEqual(0);
    });
    
    afterEach(() => {
        useSelector.mockClear();
        searchPeroopleComponent.unmount()
    });
});
describe('Test SearchPeoople component with props Placeholder', () => {
    const mockCallBack = jest.fn(()=> true);
    let searchPeroopleComponent;
    beforeEach(() => {
        useSelector.mockImplementation(callback => {
            return callback({
                family: {
                    options: [
                        {
                            "id": "2a719cc0-6c01-4dea-90c3-5ede17ced68b",
                            "name": "Austria-Hungary (historical)"
                        }
                    ],
                    optionLoaing: false
                }
            })
        });
         searchPeroopleComponent =  mount(
        <Formik
            initialValues={{ location: null
            }}
            enableReinitialize={true}
          >
            <Field
              id="location"
              name="location"
              placeholder={"Location"}
              component={FWSearchLocation}
              freeSolo={false}
              closeIconDisable={true}
              highlight={true}
            />
       </Formik>)
    })
    it('match search Placeholder', () => {
        expect(searchPeroopleComponent.props().children.props.placeholder).toEqual("Location");
    });
    it('handle Input Change',  async () => {
        const searchInput = searchPeroopleComponent.find("input");
        await act(async () => {
            const event = { target: {name: "search", value: "Austria"}};
            searchInput.simulate('change', event);
        });
        expect(mockCallBack.mock.calls.length).toEqual(0);
    });
    it('handle Input Change with id and with searchType',  async () => {
        const searchInput = searchPeroopleComponent.find("Typeahead");
        await act(async () => {
            searchInput.prop('handleChange')({
                "id": "2a719cc0-6c01-4dea-90c3-5ede17ced681",
                "name": "Austria-Hungary (historical)"
            });
        });
        await act(async () => {
            jest.runOnlyPendingTimers();
        });
        expect(mockCallBack.mock.calls.length).toEqual(0);
    });
    afterEach(() => {
        useSelector.mockClear();
        searchPeroopleComponent.unmount()
      });
});