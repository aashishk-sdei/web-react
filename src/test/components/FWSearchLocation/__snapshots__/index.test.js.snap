// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test SearchPeoople component with props match snapshot 1`] = `ReactWrapper {}`;
